rpc/src/server/service.rs

```diff
@@ -406,6 +447,213 @@ mod fns {
+
+        // get the whole context
+        let context = {
+            let reader = list.read().unwrap();
+            if let Ok(Some(ctx)) = reader.get(level) {
+                ctx
+            } else {
+                bail!("Context not found")
+            }
+        };
+
+        let roll_data = context.clone();
+        // get all the relevant data out of the context database
+        let data: HashMap<String, Bucket<Vec<u8>>> = context.into_iter()
+            .filter(|(k, _)| k.contains("data/rolls/owner/current") || k.contains("/successor") || k.contains("/random_seed"))
+            .collect();
+
+        // get the roll_lists out of the context storage
+        // roll_lists are the beginning of a linked list of the rolls pointing to the data/rolls/owner/current/*/*/<first_roll>
+        // This seems a little redundant... but with the above solution we only traverse the whole context only once
+        let roll_lists: HashMap<String, Bucket<Vec<u8>>> = roll_data.into_iter()
+            .filter(|(k, _)| k.contains("roll_list"))
+            .collect();
+            
+        let mut roll_owners: HashMap<i64, String> = HashMap::new();
+
+        for key in roll_lists.keys() {
+            // extract the delegate address (public key hash <pkh>) from the key
+            // let public_key_hash: String = key.split('/').collect::<Vec<_>>()[9].to_string(); //[9]
+            let public_key_hash: String = match key.split('/').collect::<Vec<_>>().get(ADDRESS_POSITION) {
+                Some(v) => v.to_string(),
+                None => bail!("Address not found in key")
+            };
+
+            let contract_id = address_to_contract_id(&public_key_hash)?;
+
+            // get the first roll
+            let mut next_roll;
+            if let Some(Bucket::Exists(roll)) = roll_lists.get(key) {
+                next_roll = roll;
+            } else {
+                bail!("No first roll for the delegate")
+            }
+
+            // fill out the roll_owners hash map using the linked list from the context storage
+            loop {
+                let roll_num = num_from_slice!(next_roll, ROLL_NUM_START, i32);
+
+                // construct the whole owner/successor key
+                // /<rol_num little-endian 1st byte>/<roll_num little-endian 2nd byte>/<roll_num in decimal>
+                // Note: this will work up to 0xffff rolls, not sure how exactly they implement rolls over this number with this key format
+                let owner_key = format!("data/rolls/owner/current/{}/{}/{}", next_roll[3], next_roll[2], roll_num);
+                let successor_key = format!("data/rolls/index/{}/{}/{}/successor", next_roll[3], next_roll[2], roll_num);
+            
+                if let Some(Bucket::Exists(_roll)) = data.get(&owner_key) {
+                    roll_owners.insert(roll_num.into(), contract_id.clone());
+                } else {
+                    bail!("Roll owner key not found for key: {}", &owner_key)
```

Comment:
there can be also Bucket::Deleted in owner_key, check for it or put break instead of bail